test/model/src/include/model/core.h

```diff
@@ -0,0 +1,437 @@
+ * software under copyright law.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+ * OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+ * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#ifndef MODEL_CORE_H
+#define MODEL_CORE_H
+
+#include <deque>
+#include <limits>
+#include <map>
+#include <mutex>
+#include <optional>
+#include <string>
+
+/* Redefine important WiredTiger internal constants, if they are not already available. */
+
+/*
+ * WT_TS_MAX --
+ *     The maximum timestamp, typically used in reads where we would like to get the latest value.
+ */
+#ifndef WT_TS_MAX
+#define WT_TS_MAX UINT64_MAX
+#endif
+
+/*
+ * WT_TS_NONE --
+ *     No timestamp, e.g., when performing a non-timestamped update.
+ */
+#ifndef WT_TS_NONE
+#define WT_TS_NONE 0
+#endif
+
+namespace model {
+
+/*
+ * timestamp_t --
+ *     The timestamp. This is the model's equivalent of wt_timestamp_t.
+ */
+using timestamp_t = uint64_t;
+
+/* Verify that WiredTiger constants match our expectations for the model's timestamp type. */
+static_assert(WT_TS_MAX == std::numeric_limits<timestamp_t>::max());
+static_assert(WT_TS_NONE == std::numeric_limits<timestamp_t>::min());
+
+/*
+ * NONE_STRING --
+ *     The "string" to print in place of NONE.
+ */
+extern const std::string NONE_STRING;
+
+/*
+ * data_value --
+ *     The data value stored in the model used for keys and values. We use a generic class, rather
+ *     than a specific type such as std::string, to give us flexibility to change data types in the
+ *     future, e.g., if this becomes necessary to explore additional code paths. This class is
+ *     intended to parallel WiredTiger's WT_ITEM, which supports multiple data types, plus the
+ *     ability to specify a NONE value to simplify modeling deleted data.
+ */
+class data_value {
+
+public:
+    /*
+     * data_value::data_value --
+     *     Create a new instance.
+     */
+    inline data_value(const char *data) : _data(data) {}
+
+    /*
+     * data_value::data_value --
+     *     Create a new instance.
+     */
+    inline data_value(const std::string &data) noexcept : _data(data) {}
+
+    /*
+     * data_value::data_value --
+     *     Create a new instance.
+     */
+    inline data_value(const std::string &&data) noexcept : _data(std::move(data)) {}
+
+    /*
+     * data_value::create_none --
+     *     Create an instance of a "None" value.
+     */
+    inline static data_value
+    create_none() noexcept
+    {
+        return data_value();
+    }
+
+    /*
+     * data_value::as_string --
+     *     Return the data value as a human-readable string (e.g., for printing).
+     */
+    inline const std::string &
```

Comment:
C++17 has `std::string_view` to support this use-case.

Reply:
Thanks for the pointer—`std::string_view` indeed looks useful. But actually, I ended up deleting this function. My intention was to have something along the lines of `Object.toString()` in Java, but as I started moving away from supporting just strings in `data_value`, this became fairly problematic. I see that the standard C++ way for human-readable output is overloading `std::ostream`'s `operator<<`, so I did that instead.